javascript/ql/src/semmle/javascript/frameworks/Electron.qll

```diff
@@ -6,31 +6,466 @@ module Electron {
+   * A data flow node that is sent as an asynchronous Electron IPC reply from a renderer process.
+   */
+  class IPCRendererAsyncReplyMessage extends IPCRendererReplyMessage {
+    DataFlow::Node channel;
+    
+    IPCRendererAsyncReplyMessage() {
+      exists(IPCRendererCallback callback, DataFlow::MethodCallNode mc |
+        callback.getParameter(0).getAPropertyRead("sender").flowsTo(mc.getReceiver()) and
+        this = mc.getArgument(1) and
+        mc.getCalleeName() = "send" and
+        channel = mc.getArgument(0)
+      )
+    }
+    
+    override DataFlow::Node getChannel() {
+      result = channel
+    }
+    
+    override string getChannelValue() {
+      result = channel.asExpr().getStringValue()
+    }
+  }
+  
+
+  /**
+   * A data flow node that is sent as a synchronous Electron IPC message from a renderer process.
+   */
+  class IPCRendererSyncReplyMessage extends IPCRendererReplyMessage {
+    DataFlow::Node channel;
+    
+    IPCRendererSyncReplyMessage() {
+      exists(IPCRendererCallback callback, DataFlow::PropWrite write |
+        callback.getParameter(0).flowsTo(write.getBase()) and
+        this = write.getRhs() and
+        write.getPropertyName() = "returnValue" and
+        channel = callback.getChannel()
+      )
+    }
+    
+    override DataFlow::Node getChannel() {
+      result = channel
+    }
+    
+    override string getChannelValue() {
+      result = channel.asExpr().getStringValue()
+    }
+  }
+  
+  /**
+   * A data flow node that is a call to a synchronous Electron IPC method in a renderer process.
+   */
+  class IPCRendererSendSync extends DataFlow::MethodCallNode {
+    DataFlow::Node channel;
+    
+    IPCRendererSendSync() {
+      exists(IPCRenderer ipcRenderer |
+        ipcRenderer.flowsTo(this.getReceiver()) and
+        this.getCalleeName() = "sendSync" and
+        this.getArgument(0) = channel
+      )
+    }
+    
+    DataFlow::Node getChannel() {
+      result = channel
+    }
+    
+    string getChannelValue() {
+      result = channel.asExpr().getStringValue()
+    }
+  }
+
+  
+  /**
+   * A data flow node that is sent as an asynchronous Electron IPC message from the main process via a `webContents` object.
+   */
+  class WebContentsSendMessage extends IPCMainMessage {
+    DataFlow::Node channel;
+    
+    WebContentsSendMessage() {
+      exists(WebContents wc, DataFlow::MethodCallNode mc |
+        wc.flowsTo(mc.getReceiver()) and
+        this = mc.getArgument(1) and
+        channel = mc.getArgument(0) and
+        mc.getCalleeName() = "send"
+      )
+    }
+    
+    override DataFlow::Node getChannel() {
+      result = channel
+    }
+    
+    override string getChannelValue() {
+      result = channel.asExpr().getStringValue()
+    }
+  }
+  /**
+   * Holds if `pred` flows to `succ` via the Electron IPC channel `channel`, as determined from local string values.
+   */
+  predicate ipcSimpleFlowStep(DataFlow::Node pred, DataFlow::Node succ, string channel) {
```

Comment:
Please add inline comments explaining the individual cases.

Reply:
Added